cardano-crypto-class/src/Cardano/Crypto/DSIGN/SchnorrSecp256k1.hs

```diff
@@ -0,0 +1,296 @@
+                  algorithmNameDSIGN,
+                  deriveVerKeyDSIGN, 
+                  signDSIGN, 
+                  verifyDSIGN, 
+                  genKeyDSIGN, 
+                  rawSerialiseSigDSIGN,
+                  Signable, 
+                  rawSerialiseVerKeyDSIGN, 
+                  rawSerialiseSignKeyDSIGN, 
+                  rawDeserialiseVerKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN, 
+                  rawDeserialiseSigDSIGN), 
+  encodeVerKeyDSIGN, 
+  encodedVerKeyDSIGNSizeExpr, 
+  decodeVerKeyDSIGN, 
+  encodeSignKeyDSIGN, 
+  encodedSignKeyDESIGNSizeExpr, 
+  decodeSignKeyDSIGN, 
+  encodeSigDSIGN, 
+  encodedSigDSIGNSizeExpr, 
+  decodeSigDSIGN
+  )
+import Cardano.Crypto.Util (SignableRepresentation (getSignableRepresentation))
+
+data SchnorrSecp256k1DSIGN
+
+instance DSIGNAlgorithm SchnorrSecp256k1DSIGN where
+  type SeedSizeDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeSigDSIGN SchnorrSecp256k1DSIGN = 64
+  type SizeSignKeyDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeVerKeyDSIGN SchnorrSecp256k1DSIGN = 64
+  type Signable SchnorrSecp256k1DSIGN = SignableRepresentation
+  newtype VerKeyDSIGN SchnorrSecp256k1DSIGN = 
+    VerKeySchnorr256k1 (ForeignPtr SECP256k1XOnlyPubKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "VerKeySchnorr256k1" (ForeignPtr SECP256k1XOnlyPubKey)
+        )
+  newtype SignKeyDSIGN SchnorrSecp256k1DSIGN = 
+    SignKeySchnorr256k1 (ForeignPtr SECP256k1SecKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SignKeySchnorr256k1" (ForeignPtr SECP256k1SecKey)
+        )
+  newtype SigDSIGN SchnorrSecp256k1DSIGN = 
+    SigSchnorr256k1 (ForeignPtr SECP256k1SchnorrSig)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SigSchnorr256k1" (ForeignPtr SECP256k1SchnorrSig)
+        )
+  algorithmNameDSIGN _ = "schnorr-secp256k1"
+  {-# NOINLINE deriveVerKeyDSIGN #-}
+  deriveVerKeyDSIGN (SignKeySchnorr256k1 fp) = 
+    unsafeDupablePerformIO . withForeignPtr fp $ \skp -> do
+      let skp' :: Ptr CUChar = castPtr skp
+      allocaBytes 96 $ \kpp -> do
+        res <- secpKeyPairCreate ctxPtr kpp skp'
+        when (res /= 1) (error "deriveVerKeyDSIGN: Failed to create keypair")
+        xonlyFP <- mallocForeignPtrBytes 64
+        res' <- withForeignPtr xonlyFP $ \xonlyp -> 
+          secpKeyPairXOnlyPub ctxPtr xonlyp nullPtr kpp
+        when (res' /= 1) (error "deriveVerKeyDSIGN: could not extract xonly pubkey")
+        pure . VerKeySchnorr256k1 $ xonlyFP
+  {-# NOINLINE signDSIGN #-}
+  signDSIGN () msg (SignKeySchnorr256k1 skfp) = 
+    unsafeDupablePerformIO . withForeignPtr skfp $ \skp -> do
+      let bs = getSignableRepresentation msg
+      let skp' :: Ptr CUChar = castPtr skp
+      allocaBytes 96 $ \kpp -> do
+        res <- secpKeyPairCreate ctxPtr kpp skp'
+        when (res /= 1) (error "signDSIGN: Failed to create keypair")
+        sigFP <- mallocForeignPtrBytes 64
+        let (msgFP, msgOff, msgLen) = toForeignPtr bs
```

Comment:
I think you could replace many (all?) the uses of `toForeignPointer` with `useAsCString(Len)`, which doesn't require the bytestring `Internal` module.